concurrency: datastructure changes to enable multiple lock strengths

arulajmani opened the pull request:
Prior to this patch, any unreplicated lock was assumed to be held with
lock strength Exclusive. This patch enables us to represent other lock
strengths in the lock table, which will allow us to support shared locks
in the near future.

There's a small functional change in this commit. Previously, we'd track
all sequence numbers an unreplicated lock is held at in the lock table.
As described in the shared locks RFC, this tracking is superflous --
instead, it's sufficient to only track the lowest sequence number (that
hasn't been rolled back) with which a lock is held. We do so now.

Closes https://github.com/cockroachdb/cockroach/issues/102270

Epic: none

Release note: None

arulajmani committed:
concurrency: pull clearing {,un}replicated holder info behind a function

In an upcoming commit, the zero value of unreplicatedLockHolderInfo will
no longer be a viable default. This will come in handy then.

The change to replicatedLockHolderInfo is made to preserve consistency
across these two types.

Epic: none

Release note: None

arulajmani committed:
concurrency: stop storing sequence number history for unreplicated locks

Previously, we would store all sequence numbers (that hadn't been rolled
back) that a lock was acquired at. This tracking was done to ensure that
locks were correctly rolled back when savepoints were rolled back.

However, as described in the shared locks RFC, this tracking is
superflous. Instead, it's sufficient to track just the lowest sequence
number (that hasn't been rolled back) at which a lock is held. This
relies on two invariants:

1. The lock table should hear about all ignored sequence numbers when
a lock is acquired/updated.
2. Savepoint rollback should always ignore the tail of sequence numbers
(as opposed to arbitary ones).

Both these invariants hold, which allows us to make this change now.

Release note: None

arulajmani committed:
concurrency: fix test

A single rollback operation can only rollback the tail of sequence
numbers -- random sequence numbers cannot be rolled back. This fixes
a test which was rolling back random sequence numbers.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/107538)
<!-- Reviewable:end -->

nvb reviewed (commented):
> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 34 of 34 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107538)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`-- commits` line 30 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaJHceR7tt-Kx7oFA4K):*
We discussed pulling this into a separate commit. Let me know if you decide to do that.

EDIT: after reviewing the change, I do think we should do this. It will result in a bit of additional test churn, but it will be helpful to separate out the refactor that adds multiple locking strengths from the change that begins making stronger assumptions about uses of the lock table.
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 1739 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaJH6PLC7QjXhCfJbGw:-NaJH6PLC7QjXhCfJbGx:bipi47q) ([raw file](https://github.com/cockroachdb/cockroach/blob/712e27024a4385738a4922cb6a9556963f90d219/pkg/kv/kvserver/concurrency/lock_table.go#L1739)):*
> ```Go
> 	l.holder.txn = nil
> 	l.holder.startTime = time.Time{}
> 	l.holder.replicatedInfo = replicatedLockHolderInfo{}
> ```

Consider adding a `clear` method for `replicatedLockHolderInfo` as well. Code that reintroduces any instances of `l.holder.unreplicatedInfo = unreplicatedLockHolderInfo{}` will cause bugs, and making the two structs behave more symmetrically might prevent those kinds of mistakes.
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 926 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaJI5te-SPlHI4QMzC4:-NaJI5te-SPlHI4QMzC5:b-n9wz3o) ([raw file](https://github.com/cockroachdb/cockroach/blob/f56e347cc45a44438ea2e21b9882c300e45cd63a/pkg/kv/kvserver/concurrency/lock_table.go#L926)):*
> ```Go
> // Information about a lock holder for unreplicated locks.
> type unreplicatedLockHolderInfo struct {
> 	// infos tracks whether the lock is held with a particular strength; if it is,
> ```

Is "infos" the right name for this? Did you consider "strengths"?
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 933 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaJKR-k7HYGPXka5r3T:-NaJKR-k7HYGPXka5r3U:b-3bnu13) ([raw file](https://github.com/cockroachdb/cockroach/blob/f56e347cc45a44438ea2e21b9882c300e45cd63a/pkg/kv/kvserver/concurrency/lock_table.go#L933)):*
> ```Go
> 	// NB: Intents cannot be held/acquired in unreplicated fashion; thus the
> 	// highest lock strength for unreplicated locks is Exclusive.
> 	infos [lock.Exclusive + 1]enginepb.TxnSeq
> ```

There's space in this array for `lock.None` and `lock.Update`, neither of which can actually be stored in here. What do you think about a bit of indirection to slim this down to its minimum necessary size and ensure that incorrect usage results in a runtime error. Something like:
```go
var unreplicatedHolderStrengths = [...]lock.Strength{lock.Shared, lock.Exclusive}
var unreplicatedHolderStrengthToIndexMap = func() [lock.MaxStrength]int {
	var m [lock.MaxStrength]int
	// Initialize all to -1.
	for str := range m {
		m[str] = -1
	}
	// Set the indices of the valid strengths.
	for i, str := range unreplicatedHolderStrengths {
		m[str] = i
	}
	return m
}()
var unreplicatedHolderIndexToStrengthMap = func() [len(unreplicatedHolderStrengths)]lock.Strength {
	var m [len(unreplicatedHolderStrengths)]lock.Strength
	for i, str := range unreplicatedHolderStrengths {
		m[i] = str
	}
	return m
}()

type unreplicatedLockHolderInfo struct {
	/// ...
	infos [len(unreplicatedHolderStrengths)]enginepb.TxnSeq
	/// ...
}

func (ulh *unreplicatedLockHolderInfo) held(str lock.Strength) bool {
	return ulh.infos[unreplicatedHolderStrengthToIndexMap[str]] != -1
}
```
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 941 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaJIZj48iitl4IUIVTX:-NaJIZj48iitl4IUIVTY:bnimiov) ([raw file](https://github.com/cockroachdb/cockroach/blob/f56e347cc45a44438ea2e21b9882c300e45cd63a/pkg/kv/kvserver/concurrency/lock_table.go#L941)):*
> ```Go
> // init initializes an unreplicatedLockHolderInfo struct.
> func (ulh *unreplicatedLockHolderInfo) init() {
> 	for str := lock.None; str <= lock.Exclusive; str++ {
> ```

nit: is this the same as `for str := range ulh.infos {`?
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 948 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaJIWCX5od_Q1Z5TUHa:-NaJIWCX5od_Q1Z5TUHb:b729bgw) ([raw file](https://github.com/cockroachdb/cockroach/blob/f56e347cc45a44438ea2e21b9882c300e45cd63a/pkg/kv/kvserver/concurrency/lock_table.go#L948)):*
> ```Go
> // clear removes previously tracked unreplicated lock holder information.
> func (ulh *unreplicatedLockHolderInfo) clear() {
> 	for str := lock.None; str <= lock.Exclusive; str++ {
> ```

There's some duplication here. Should we have `clear` reset `ulh` and then call `init`?
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 957 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaJIkdBCQXQfaVw_9YO:-NaJIkdBCQXQfaVw_9YP:bkn40bb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f56e347cc45a44438ea2e21b9882c300e45cd63a/pkg/kv/kvserver/concurrency/lock_table.go#L957)):*
> ```Go
> // State specific to the previous epoch is cleared out.
> func (ulh *unreplicatedLockHolderInfo) epochBumped() {
> 	for str := lock.None; str <= lock.Exclusive; str++ {
> ```

Should we just call `init`? Or maybe a `resetStrengths` method that both `init` and `epochBumped` call?
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 969 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaJJ7GN8iXtSYO41xqh:-NaJJ7GN8iXtSYO41xqi:b5c06bk) ([raw file](https://github.com/cockroachdb/cockroach/blob/f56e347cc45a44438ea2e21b9882c300e45cd63a/pkg/kv/kvserver/concurrency/lock_table.go#L969)):*
> ```Go
> func (ulh *unreplicatedLockHolderInfo) acquire(str lock.Strength, seqNum enginepb.TxnSeq) {
> 	if !ulh.held(str) ||
> 		// ... OR we've been made aware of an acquisition at a lower sequence number
> ```

nit: punctuation
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 969 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaJJR8zAteBYICWFYhw:-NaJJR8zAteBYICWFYhx:b-7s13wr) ([raw file](https://github.com/cockroachdb/cockroach/blob/f56e347cc45a44438ea2e21b9882c300e45cd63a/pkg/kv/kvserver/concurrency/lock_table.go#L969)):*
> ```Go
> func (ulh *unreplicatedLockHolderInfo) acquire(str lock.Strength, seqNum enginepb.TxnSeq) {
> 	if !ulh.held(str) ||
> 		// ... OR we've been made aware of an acquisition at a lower sequence number
> ```

Can this case happen in practice? It seems to break the assumption we are making about ignored sequence numbers. If this case can happen, then what would happen if the earlier acquisition with the higher sequence number also rolled back this later, lower sequence number?
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 988 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaK3_Za2MYGVtB88ICU:-NaK3_Za2MYGVtB88ICV:b-sgj38m) ([raw file](https://github.com/cockroachdb/cockroach/blob/f56e347cc45a44438ea2e21b9882c300e45cd63a/pkg/kv/kvserver/concurrency/lock_table.go#L988)):*
> ```Go
> 		return
> 	}
> 	for str := lock.None; str <= lock.Exclusive; str++ {
> ```

Same point here and below about `for i := range ulh.infos` being less error-prone for iterating over all elements in the array.
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 2382 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaK4tzC41rYlMOGBNFY:-NaK4tzC41rYlMOGBNFZ:b-n1gzrr) ([raw file](https://github.com/cockroachdb/cockroach/blob/f56e347cc45a44438ea2e21b9882c300e45cd63a/pkg/kv/kvserver/concurrency/lock_table.go#L2382)):*
> ```Go
> 			switch {
> 			case l.holder.txn.Epoch < acq.Txn.Epoch: // at a higher epoch
> 				// Clear sequence number tracking from the older epoch.
> ```

"but keep the timestamp", right? That's why this can't just call `clear`?
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 2812 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaK5Pp20CoNCkFTiZxN:-NaK5Pp20CoNCkFTiZxO:b-fsrt2n) ([raw file](https://github.com/cockroachdb/cockroach/blob/f56e347cc45a44438ea2e21b9882c300e45cd63a/pkg/kv/kvserver/concurrency/lock_table.go#L2812)):*
> ```Go
> 			// numbers.
> 			held := false
> 			for str := lock.None; str <= lock.Exclusive; str++ {
> ```

Is this the same as `isEmpty()` if we get rid of the assertion?
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 3635 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaK65MMBZcCBDJ1QJQC:-NaK65MMBZcCBDJ1QJQD:bwwdi40) ([raw file](https://github.com/cockroachdb/cockroach/blob/f56e347cc45a44438ea2e21b9882c300e45cd63a/pkg/kv/kvserver/concurrency/lock_table.go#L3635)):*
> ```Go
> 
> // QueryLockTableState implements the lockTable interface.
> func (t *lockTableImpl) QueryLockTableState(
> ```

This reminds me that we'll want to update `QueryLocksRequest` to return the strength(s) of the locks that are held, so that they can be used to populate `crdb_internal.cluster_locks`. We currently just assume the locks is held with an Exclusive strength [here](https://github.com/cockroachdb/cockroach/blob/bac4c7949a2621da62da3bdb31924a2e7c51d5bc/pkg/sql/crdb_internal.go#L7658). Do you mind opening an issue for that so we can track it going forward? It might make a good intern project, unless you'd prefer to just do it yourself.


<!-- Sent from Reviewable.io -->

arulajmani committed:
concurrency: datastructure changes to enable multiple lock strengths

Prior to this patch, any unreplicated lock was assumed to be held with
lock strength Exclusive. This patch enables us to represent other lock
strengths in the lock table, which will allow us to support shared locks
in the near future.

Closes #102270

Release note: None

arulajmani reviewed (commented):
TFTR, RFAL! 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-:-NaOenql0u_w0XpDmNtt:bmu8o4n)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[`-- commits` line 30 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaJHceR7tt-Kx7oFA4K):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We discussed pulling this into a separate commit. Let me know if you decide to do that.

EDIT: after reviewing the change, I do think we should do this. It will result in a bit of additional test churn, but it will be helpful to separate out the refactor that adds multiple locking strengths from the change that begins making stronger assumptions about uses of the lock table.
</blockquote></details>

Pulled it out into its own commit. Ended up making the formatting such so that there's no test churn between commits 3 and 4. 
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 1739 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaJH6PLC7QjXhCfJbGw:-NaMo4E6AC5mMFyEPZFs:b635ybr) ([raw file](https://github.com/cockroachdb/cockroach/blob/712e27024a4385738a4922cb6a9556963f90d219/pkg/kv/kvserver/concurrency/lock_table.go#L1739)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Consider adding a `clear` method for `replicatedLockHolderInfo` as well. Code that reintroduces any instances of `l.holder.unreplicatedInfo = unreplicatedLockHolderInfo{}` will cause bugs, and making the two structs behave more symmetrically might prevent those kinds of mistakes.
</blockquote></details>

Makes sense; done. 
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 926 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaJI5te-SPlHI4QMzC4:-NaNvb39EEoxJZ1VXIAH:b9cbb95) ([raw file](https://github.com/cockroachdb/cockroach/blob/f56e347cc45a44438ea2e21b9882c300e45cd63a/pkg/kv/kvserver/concurrency/lock_table.go#L926)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is "infos" the right name for this? Did you consider "strengths"?
</blockquote></details>

Switched to strengths.
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 933 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaJKR-k7HYGPXka5r3T:-NaOZEc_7yD42jHjK67u:boo7h1h) ([raw file](https://github.com/cockroachdb/cockroach/blob/f56e347cc45a44438ea2e21b9882c300e45cd63a/pkg/kv/kvserver/concurrency/lock_table.go#L933)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

There's space in this array for `lock.None` and `lock.Update`, neither of which can actually be stored in here. What do you think about a bit of indirection to slim this down to its minimum necessary size and ensure that incorrect usage results in a runtime error. Something like:
```go
var unreplicatedHolderStrengths = [...]lock.Strength{lock.Shared, lock.Exclusive}
var unreplicatedHolderStrengthToIndexMap = func() [lock.MaxStrength]int {
	var m [lock.MaxStrength]int
	// Initialize all to -1.
	for str := range m {
		m[str] = -1
	}
	// Set the indices of the valid strengths.
	for i, str := range unreplicatedHolderStrengths {
		m[str] = i
	}
	return m
}()
var unreplicatedHolderIndexToStrengthMap = func() [len(unreplicatedHolderStrengths)]lock.Strength {
	var m [len(unreplicatedHolderStrengths)]lock.Strength
	for i, str := range unreplicatedHolderStrengths {
		m[i] = str
	}
	return m
}()

type unreplicatedLockHolderInfo struct {
	/// ...
	infos [len(unreplicatedHolderStrengths)]enginepb.TxnSeq
	/// ...
}

func (ulh *unreplicatedLockHolderInfo) held(str lock.Strength) bool {
	return ulh.infos[unreplicatedHolderStrengthToIndexMap[str]] != -1
}
```
</blockquote></details>

This is neat, made the change. 
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 941 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaJIZj48iitl4IUIVTX:-NaNvWwH3JJNyU7sHs6o:bss6z5a) ([raw file](https://github.com/cockroachdb/cockroach/blob/f56e347cc45a44438ea2e21b9882c300e45cd63a/pkg/kv/kvserver/concurrency/lock_table.go#L941)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: is this the same as `for str := range ulh.infos {`?
</blockquote></details>

Yup, made the change in `resetStrenghts` now that this calls into that.
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 948 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaJIWCX5od_Q1Z5TUHa:-NaNl9cB77p2Jp1TbTpY:bt0o6em) ([raw file](https://github.com/cockroachdb/cockroach/blob/f56e347cc45a44438ea2e21b9882c300e45cd63a/pkg/kv/kvserver/concurrency/lock_table.go#L948)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

There's some duplication here. Should we have `clear` reset `ulh` and then call `init`?
</blockquote></details>

Ended up calling `resetStrengths` now that we have that to avoid the duplication.
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 957 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaJIkdBCQXQfaVw_9YO:-NaNlEzk5kjPVqedsO-N:b-bjlqlm) ([raw file](https://github.com/cockroachdb/cockroach/blob/f56e347cc45a44438ea2e21b9882c300e45cd63a/pkg/kv/kvserver/concurrency/lock_table.go#L957)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we just call `init`? Or maybe a `resetStrengths` method that both `init` and `epochBumped` call?
</blockquote></details>

Chose to add `resetStrengths`.
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 969 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaJJR8zAteBYICWFYhw:-NaOZQ7OAHEcCs5aYxQ4:b-bjdbp5) ([raw file](https://github.com/cockroachdb/cockroach/blob/f56e347cc45a44438ea2e21b9882c300e45cd63a/pkg/kv/kvserver/concurrency/lock_table.go#L969)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Can this case happen in practice? It seems to break the assumption we are making about ignored sequence numbers. If this case can happen, then what would happen if the earlier acquisition with the higher sequence number also rolled back this later, lower sequence number?
</blockquote></details>

Like we spoke about offline, this shouldn't be possible in practice -- by extension, the assumptions we're making in just tracking the minimum (non-rolled back) sequence number a lock was acquired at are correct. 
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 2382 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaK4tzC41rYlMOGBNFY:-NaNhuob2TV_yT9prkCu:bizb7t4) ([raw file](https://github.com/cockroachdb/cockroach/blob/f56e347cc45a44438ea2e21b9882c300e45cd63a/pkg/kv/kvserver/concurrency/lock_table.go#L2382)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

"but keep the timestamp", right? That's why this can't just call `clear`?
</blockquote></details>

Yeah. Ended up improving the comment on `epochBumped` and removing this comment entirely -- this seemed unnecessary now that we have a well named function.
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 2812 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaK5Pp20CoNCkFTiZxN:-NaOZevo4MLwBVRj31pv:b2iutyo) ([raw file](https://github.com/cockroachdb/cockroach/blob/f56e347cc45a44438ea2e21b9882c300e45cd63a/pkg/kv/kvserver/concurrency/lock_table.go#L2812)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is this the same as `isEmpty()` if we get rid of the assertion?
</blockquote></details>

Yeah, it is. Were you suggesting we pull out the non-assertion bit into its own function? Kept this in-lined for now, as this is the only place where an unreplicated lock will transition from unheld to held. 
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 3635 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaK65MMBZcCBDJ1QJQC:-NaOdLdx7ccdOwnVkmN8:bm3onzt) ([raw file](https://github.com/cockroachdb/cockroach/blob/f56e347cc45a44438ea2e21b9882c300e45cd63a/pkg/kv/kvserver/concurrency/lock_table.go#L3635)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This reminds me that we'll want to update `QueryLocksRequest` to return the strength(s) of the locks that are held, so that they can be used to populate `crdb_internal.cluster_locks`. We currently just assume the locks is held with an Exclusive strength [here](https://github.com/cockroachdb/cockroach/blob/bac4c7949a2621da62da3bdb31924a2e7c51d5bc/pkg/sql/crdb_internal.go#L7658). Do you mind opening an issue for that so we can track it going forward? It might make a good intern project, unless you'd prefer to just do it yourself.
</blockquote></details>

Filing; will link once it's done. 


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107538)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 3635 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaK65MMBZcCBDJ1QJQC:-NaOgVV724ndo7XJKQuh:b-ijrdc7) ([raw file](https://github.com/cockroachdb/cockroach/blob/f56e347cc45a44438ea2e21b9882c300e45cd63a/pkg/kv/kvserver/concurrency/lock_table.go#L3635)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Filing; will link once it's done. 
</blockquote></details>

https://github.com/cockroachdb/cockroach/issues/107766


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 34 of 34 files at r4, 1 of 1 files at r5, 32 of 34 files at r6, 2 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107538)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 988 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaK3_Za2MYGVtB88ICU:-NaPsKkP3eGwgq-Q4cJR:b-x8ep5g) ([raw file](https://github.com/cockroachdb/cockroach/blob/f56e347cc45a44438ea2e21b9882c300e45cd63a/pkg/kv/kvserver/concurrency/lock_table.go#L988)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Same point here and below about `for i := range ulh.infos` being less error-prone for iterating over all elements in the array.
</blockquote></details>

We now have a few of `for str := range ulh.strengths {`. Consider renaming `str` to `strIdx` to avoid readers confusing `str` with the `lock.Strength`.

Also, in this case can others where you keep accessing the array, consider pulling out the `minSeqNum` directly in the loop like `for strIdx, minSeqNum := range ulh.strengths {`
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 3635 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaK65MMBZcCBDJ1QJQC:-NaPtuw2EBwDVUzEv3XK:b-ucfknh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f56e347cc45a44438ea2e21b9882c300e45cd63a/pkg/kv/kvserver/concurrency/lock_table.go#L3635)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

https://github.com/cockroachdb/cockroach/issues/107766
</blockquote></details>

Thanks!
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 983 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaPtPLHAAcRFGtCRr6q:-NaPtPLHAAcRFGtCRr6r:b-9iimwe) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef1ab7d43bd0baff0cc91c74de9f5db34b7c1eb6/pkg/kv/kvserver/concurrency/lock_table.go#L983)):*
> ```Go
> 
> func (ulh *unreplicatedLockHolderInfo) resetStrengths() {
> 	for str := range ulh.strengths {
> ```

Same comment as down below about `strIdx`.
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 1063 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaPtVMJDW-9k6_83CFE:-NaPtVMJDW-9k6_83CFF:b-6tanb8) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef1ab7d43bd0baff0cc91c74de9f5db34b7c1eb6/pkg/kv/kvserver/concurrency/lock_table.go#L1063)):*
> ```Go
> 			"(str: %s seq: %d)",
> 			redact.Safe(str),
> 			redact.Safe(ulh.strengths[unreplicatedLockHolderStrengthToIndexMap[str]]),
> ```

Would it be worthwhile to add a `minSeqNum(str lock.Strength) enginepb.TxnSeq` method to hide away this mapping? You could even call that in `held` and in parts of `acquire` so that `unreplicatedLockHolderStrengthToIndexMap` is used in even fewer places.
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 1844 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaPt8hQCBcwcvRWofa-:-NaPt8hQCBcwcvRWofa0:b-dq5pri) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef1ab7d43bd0baff0cc91c74de9f5db34b7c1eb6/pkg/kv/kvserver/concurrency/lock_table.go#L1844)):*
> ```Go
> 		return lock.MakeModeIntent(lockHolderTS)
> 	}
> 	for _, str := range unreplicatedHolderStrengths {
> ```

Do we need to check that the lock is held at this strength?
___
*[`pkg/kv/kvserver/concurrency/lock_table_test.go` line 1891 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaPuCHGB8vLC5Or9Jmr:-NaPuCHGB8vLC5Or9Jms:biab0id) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef1ab7d43bd0baff0cc91c74de9f5db34b7c1eb6/pkg/kv/kvserver/concurrency/lock_table_test.go#L1891)):*
> ```Go
> 	l.holder.unreplicatedInfo.init()
> 	l.holder.unreplicatedInfo.ts = hlc.Timestamp{WallTime: 123, Logical: 7}
> 	l.holder.unreplicatedInfo.strengths[unreplicatedLockHolderStrengthToIndexMap[lock.Exclusive]] = 1
> ```

nit: can you use the `acquire` method?


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
RFAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-:-NaTjyx433pT_4mRqpLL:bb72pd0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 988 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaK3_Za2MYGVtB88ICU:-NaTk9GeEOL-N-BHo2pz:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/f56e347cc45a44438ea2e21b9882c300e45cd63a/pkg/kv/kvserver/concurrency/lock_table.go#L988)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We now have a few of `for str := range ulh.strengths {`. Consider renaming `str` to `strIdx` to avoid readers confusing `str` with the `lock.Strength`.

Also, in this case can others where you keep accessing the array, consider pulling out the `minSeqNum` directly in the loop like `for strIdx, minSeqNum := range ulh.strengths {`
</blockquote></details>

Done. 
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 1063 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaPtVMJDW-9k6_83CFE:-NaTi6He81nFl9pwLR3A:buu46be) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef1ab7d43bd0baff0cc91c74de9f5db34b7c1eb6/pkg/kv/kvserver/concurrency/lock_table.go#L1063)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Would it be worthwhile to add a `minSeqNum(str lock.Strength) enginepb.TxnSeq` method to hide away this mapping? You could even call that in `held` and in parts of `acquire` so that `unreplicatedLockHolderStrengthToIndexMap` is used in even fewer places.
</blockquote></details>

Yeah, this stuff is quite verbose otherwise. Done. 

There's just a single place we set this field and use the `unreplicatedLockHolderStrengthToIndexMap`, so I didn't bother with adding a setter.
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 1844 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaPt8hQCBcwcvRWofa-:-NaTiH4f0Is_EVa2D2o2:b-ti9czx) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef1ab7d43bd0baff0cc91c74de9f5db34b7c1eb6/pkg/kv/kvserver/concurrency/lock_table.go#L1844)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we need to check that the lock is held at this strength?
</blockquote></details>

Yeah, we do. Missed adding one while splitting r3 into two commits. Funny that no tests broke because as of this patch, there can only be exclusive locks (which must be held if we're here) and we're iterating in strength order.
___
*[`pkg/kv/kvserver/concurrency/lock_table_test.go` line 1891 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaPuCHGB8vLC5Or9Jmr:-NaTjwEO5lcGkwQyJ780:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef1ab7d43bd0baff0cc91c74de9f5db34b7c1eb6/pkg/kv/kvserver/concurrency/lock_table_test.go#L1891)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: can you use the `acquire` method?
</blockquote></details>

Done. 


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-:-NaTkahc3KdwmxFKW-X6:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 1069 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-NaTkQl26woNt8dD5Hf1:-NaTkQl26woNt8dD5Hf2:b-grqdzl) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c1b0095d96bce08f7862f6715ceed86b9d48eaf/pkg/kv/kvserver/concurrency/lock_table.go#L1069)):*
> ```Go
> }
> 
> // minSeqNumber returns the minimum sequence number the lock is held with given
> ```

nit: move this right above `held`.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
Thanks for the quick stamp! 

bors r=nvanbenschoten 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107538#-:-NaTn4s8AYeI3ML3pZpX:b2j4pgq)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten)



<!-- Sent from Reviewable.io -->
